feature: bulk transfer adapters

bogomolets-owl opened the pull request:
Hey git lfs team, hope you're doing great!

This PR introduces a new Bulk Transfer Protocol mechanism that complements the existing custom transfer mechanism to allow processing multiple files simultaneously in bulks. This feature allows different transfer planning strategies (compared to current sequential one artifact per process strategy in custom transfers) for new transfer implementations. 
This is important for repositories with many LFS objects of different file sizes. The new mechanism should allow more advanced transfer techniques like data deduplication per multiple files or multifile packing in a more easy manner compared to custom transfers.

PR contains protocol specification in docs/bulk-transfers.md and the draft implementation in bulk.go. 
Also the test infrastructure:
- test bulk lfs client with basic tests implementation
- test lfs server modification
- some basic tests

I am no good in go programming so while developing this code i used AI help. I am not entirely sure on the quality of this exact request though i tried my best to follow the patterns i saw in custom.go.

Best regards, Alexander Bogomolets.

bogomolets-owl committed:
Implement bulk adapters feature

Implement the custom adapters feature to allow custom artifacts transfer grouping via external processes.

cceneag reviewed (changes requested):
👍

bogomolets-owl committed:
Submit proposal document

This commit submits a proposal document for bulk transfers and modifies
the custom-transfers.md documentation to illustrate the protocol and
design changes

chrisd8088 commented on docs/custom-transfers.md:
```diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
+* `concurrenttransfers`: reflects the value of `lfs.concurrenttransfers`, for if
+  the transfer process wants to implement its own concurrency and wants to
+  respect this setting.
+
```
We should add a list entry here for `mode` as well:

```suggestion
* `mode`: reflects the current effective transfer mode. For the bulk protocol this
  must be set to `bulk`.

```

chrisd8088 commented on docs/custom-transfers.md:
```diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
+
+**Important**: If the number of remaining files is less than the configured
+`bulkSize`, git-lfs will automatically flush the pending transfers after a
+100ms timeout to ensure timely processing of incomplete bulks.
```
Let's say `a configurable timeout` rather than specify `100ms` exactly.

bogomolets-owl replied:
This part should be removed. New proposal text from commit 80499f0045d5a975f31ac10e9a6f40f7fc0c438a states an issue we should resulve during the proposal implementation: support explicit transfer queue flushes so adapter dont have to assume any tranfer queue states.

chrisd8088 commented on docs/custom-transfers.md:
````diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
+It looks like this:
+
+```json
+{ "event": "bulk-header", "oid": "bulk-12345-uuid", "size": 5 }
````
I think using `oid` here for a field which isn't a strict Git LFS object ID will be confusing.  May I suggest a field name like `bulk-transfer-id` instead?

chrisd8088 commented on docs/custom-transfers.md:
````diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
+like this:
+
+```json
+{ "event": "upload", "oid": "bf3e3e2af9366a3b704ae0c31de5afa64193ebabffde2091936ad2e7510bc03a", "size": 346232, "path": "/path/to/file.png", "action": { "href": "nfs://server/path", "header": { "key": "value" } } }
````
I think the bulk transfer ID should be included in these messages, for additional clarity, so I'd suggest we add the `bulk-transfer-id` field (or whatever name we choose instead for that field in the header and footer messages).

bogomolets-owl replied:
This is addressed in commit 80499f0045d5a975f31ac10e9a6f40f7fc0c438a: i opted for bid (batch identifier) field, that is persistent throughout all the messages starting from header.

chrisd8088 commented on docs/custom-transfers.md:
````diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
+like this:
+
+```json
+{ "event": "download", "oid": "22ab5f63670800cc7be06dbed816012b0dc411e774754c7579467d2536a9cf3e", "size": 21245, "action": { "href": "nfs://server/path", "header": { "key": "value" } } }
````
Again, let's include the `bulk-transfer-id` field here.

bogomolets-owl replied:
This is addressed in commit 80499f0045d5a975f31ac10e9a6f40f7fc0c438a: i opted for bid (batch identifier) field, that is persistent throughout all the messages starting from header.

chrisd8088 commented on docs/custom-transfers.md:
````diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
+It looks like this:
+
+```json
+{ "event": "progress", "oid": "bulk-12345-uuid", "bytesSoFar": 1572864, "bytesSinceLast": 524288 }
````
Again, let's use `bulk-transfer-id` in place of `oid`.

chrisd8088 commented on docs/custom-transfers.md:
```diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
+the bulk. Common strategies include:
+
+1. **Sequential Processing**: Process items one by one in the order received
+2. **Parallel Processing**: Process multiple items simultaneously using threads/workers
```
I'm not sure I'd suggest parallel processing as an option here, since IIUC that's the purpose of the existing "concurrent transfers" feature and all of its configuration settings.

For instance, if a user has set `lfs.concurrentTransfers` to a specific value, and enabled `lfs.customtransfer.<name>.concurrent`, the transfer adapter should respect those rather than proceed with "complete freedom".

bogomolets-owl replied:
This is a tricky part. You are indeed right: essential point of the whole feature is to allow processing a batch of objects as a single unit of work. This exact part might mislead custom transfer authors.
For now this part is left intact and not addressed, but proposal text now has detailed example of desired feature usage. I propose to rework this part to focus on the idea of batching the downloads. 
I will return to this part a bit later.

chrisd8088 commented on docs/custom-transfers.md:
````diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
+If there was an error during the bulk transfer, it can be sent as follows:
+
+```json
+{ "event": "bulk-complete", "oid": "bulk-12345-uuid", "error": { "code": 500, "message": "Bulk transfer failed due to network error", "retry": false } }
````
Here again, let's use `transferID` in place of `oid`.

bogomolets-owl replied:
This is addressed in commit 80499f0045d5a975f31ac10e9a6f40f7fc0c438a: i opted for bid (batch identifier) field, that is persistent throughout all the messages starting from header.

chrisd8088 commented on docs/custom-transfers.md:
````diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
+```
+
+Note: In the parallel example, file2 completes before file1, demonstrating that 
+completion order is independent of the order in which items were sent.
````
Rather than including a "parallel processing" example here, I think it would be useful to show or explain that additional bulk transfers can occur before the final `terminate` message.

The client could just start with another `bulk-header`/`bulk-footer` message pair with a series of individual object `upload` or `download` messages between them, and then process the responses, repeating as necessary before finally closing the connection with a `terminate` message.

As for the concept of parallel processing, I think we should simply explain that any parallelism is controlled by the `concurrenttransfers` settings exactly as for the `basic` mode.

chrisd8088 reviewed (commented):
Thank you so much for writing up the proposal and technical specification for this new "bulk" custom transfer adapter protocol!  I'm sorry it's taken me so long to do a thorough review.

I have some specific recommendations, mostly about the names of fields and making sure we aren't repurposing terms like `oid` and `size` for different purposes if we can avoid doing so, and indicating which fields are optional.

My main high-level recommendation is that we clarify that parallelism of bulk transfers is possible, but will be controlled by the `concurrenttransfers` settings just as it is for the `basic` mode.  I'd keep the notion of bulk transfers separate from the concept of multiple, parallel transfer queues.

Finally, I don't mention this in any specific comment, but how do you feel about including a `version` field?  We'd have to think through how the client should behave when dealing with a legacy service which didn't expect a `version` field, of course.

bogomolets-owl committed:
Rework the proposal

This commit reworks the proposal to make terminology aligned with git lfs
therms and address previous review comments.
What is done:
1. The proposal text is updated introducing naming change (bulk -> batch)
and aligning this feature with lfs batch api
2. Custom transfer protocol version is introduced and the protocol
description documentation is updated to reflect version introduction
3. Batch(bulk) protocol messages fields naming corrected to evade size
and oid field meaning redefinition in different messages
4. LFS Batch API documentation updated to reflect the proposed changes
in supporting server negotiation

bogomolets-owl commented on docs/custom-transfers.md:
````diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
+process should send a completion message for that item. 
+
+```json
+{ "event": "complete", "oid": "22ab5f63670800cc7be06dbed816012b0dc411e774754c7579467d2536a9cf3e", "path": "/tmp/downloaded/file.bin" }
````
This is addressed in commit 80499f0045d5a975f31ac10e9a6f40f7fc0c438a: i opted for bid (batch identifier) field, that is persistent throughout all the messages starting from header.

EKOVEEVOKE1 commented:
bogomolets-owl:main

chrisd8088 reviewed (commented):
Thanks very much again for all the time and effort you've put into this proposal!

You've raised some very interesting points about how the Git LFS Batch API might also be included in this new design.  However, that would increase the scope quite significantly, and given how many other services, products, and projects implement the Batch API, I think we might be better served if we can keep this proposal confined to just the custom transfer adapter.

I've tried to organize my thoughts into a few sections below.  I hope they are helpful!  We really appreciate how much thought and care you've put into this work.

#### Initial Proposal Document

One request I'd make is that the first PR we merge on this topic contain just a single file in the `docs/proposals` directory.  That could certainly be a fresh PR, so you can keep experimenting in this one!

The initial proposal document should describe the rationale for the changes, as you've already done in the `custom-transfers-bulking.md` file already.  I'd also include some JSON examples to illustrate the specific API changes that are proposed.

We can then merge that file sooner rather than later (since it's just a proposal) and make revisions and updates to it in subsequent PRs, instead of having to pack all those conversations into just this one PR.

If it's useful, please don't hesitate to include a diagram or two!  I think GitHub should be able to render MermaidJS diagrams from a Markdown file.  For instance, it helps me to think about the relationships involved in a custom transfer like this:

```mermaid
flowchart LR
    client
    api["Batch API server"]
    custom["custom transfer agent"]
    server["custom transfer server"]
    client --> api
    client --> custom
    custom --> server
```

#### Terminology

I very much appreciate how you've tried to keep our terminology consistent.  My one caution at this point is that we try to avoid using the same term in too many contexts.

For instance, I think we have to accept that the existing Batch API only has one "mode", which is "batch mode".  I don't think we want to suggest that the Batch API can be used to negotiate a non-batch mode, whether we call that "basic" or anything else.

I also think we want to leave the term "concurrency" out of this proposal as much as possible, since it really just implies parallelism, and we already support that feature, albeit with some limitations and issues (and see the next section below on that point).  On reflection, I shouldn't even have suggested the term "parallel transfers" back in my very first [comments](https://github.com/git-lfs/git-lfs/pull/6119#issuecomment-3332128952) on this PR.

I've been mulling over some possible terms for the last couple of weeks, and it seems to me that the core change you'd like to introduce is the ability for the Git LFS client to request that a custom transfer agent fetch or push a set or bundle of objects at once, rather than fetching or pushing objects individually.

So one possible term rather than "batch" or "bulk" is "bundle", or, if that doesn't suit, perhaps "set" or "group" or "pack".

The new version of the custom transfer protocol's initialization message might then include, along with a `"version": 2.0` field, a `"bundleSize"` field whose default value is assumed to be 1 for the legacy version of the protocol.  A newer Git LFS client would then send an initialization message like:

```json5
{ "event": "init", "version": 2.0,
  "operation": "download", "remote": "origin",
  "concurrent": true, "concurrenttransfers": 3,
  "bundleSize": 100 }
```

The custom transfer agent would then expect to receive messages from the Git LFS client about sets or "bundles" of Git LFS objects, with a maximum of 100 objects per bundle.  (There could of course be fewer objects in a bundle than the maximum, if there weren't 100 objects left to be transferred, or retries needed to occur, etc.)

#### Existing Implementation Issues

I thought I would take the opportunity to describe what I believe is a bit of an oversight in our existing implementation with regards to concurrency.

To be clear, I don't think we need to discuss this issue in the proposal, since that's about transferring bundles or sets of objects instead of single objects, rather than concurrency or parallelism.  (After all, even if concurrency is disabled via configuration options, a custom transfer agent could still transfer objects in bundles, so these are separate concepts.)

Nevertheless, I wanted to mention this issue because it may have occurred to you as well.

As you know, the `lfs.concurrentTransfers` option specifies the number of custom adapter processes that we [start](https://github.com/git-lfs/git-lfs/blob/1f2f7c697b5568b4bf1ed76e7ed4115dd1f0b3ba/tq/custom.go#L144) ([unless](https://github.com/git-lfs/git-lfs/blob/1f2f7c697b5568b4bf1ed76e7ed4115dd1f0b3ba/tq/custom.go#L120) the `lfs.customTransfer.<name>.concurrent` flag is set to `false`, in which case we only start [one](https://github.com/git-lfs/git-lfs/blob/1f2f7c697b5568b4bf1ed76e7ed4115dd1f0b3ba/tq/custom.go#L406) process).  We also [start](https://github.com/git-lfs/git-lfs/blob/1f2f7c697b5568b4bf1ed76e7ed4115dd1f0b3ba/tq/adapterbase.go#L96) one internal goroutine (i.e., thread) to communicate with each process.

This all occurs within the `Begin()` method of the `adapterbase` structure where we [loop](https://github.com/git-lfs/git-lfs/blob/1f2f7c697b5568b4bf1ed76e7ed4115dd1f0b3ba/tq/adapterbase.go#L91-L97) as many times as is specified by the `lfs.concurrentTransfers` option, or the value with which we've overridden it.  This loop calls the `WorkerStarting()` method of the `customAdapter` structure because we've set that structure's `transferImpl` element to [point](https://github.com/git-lfs/git-lfs/blob/1f2f7c697b5568b4bf1ed76e7ed4115dd1f0b3ba/tq/custom.go#L346) back to itself.

The `WorkerStarting()` method, in addition to starting one custom transfer agent process, then [sends](https://github.com/git-lfs/git-lfs/blob/1f2f7c697b5568b4bf1ed76e7ed4115dd1f0b3ba/tq/custom.go#L152-L155) an initialization message to the process.

That all makes sense, but where I think we have an oversight is that the initialization message [includes](https://github.com/git-lfs/git-lfs/blob/1f2f7c697b5568b4bf1ed76e7ed4115dd1f0b3ba/tq/custom.go#L68-L69) `concurrent` and `concurrenttransfers` elements, and we've [documented](https://github.com/git-lfs/git-lfs/blob/1f2f7c697b5568b4bf1ed76e7ed4115dd1f0b3ba/docs/custom-transfers.md?plain=1#L134-L138) that we include the latter specifically in case "the transfer process wants to implement its own concurrency and wants to respect this setting".

But there's no way for the process to "respect" the `concurrenttransfers` setting since we've already started that many separate processes, and each one just gets a copy of the same initialization message.

Arguably, the Git LFS client should either start a single process, pass it the concurrency settings, and let it figure out how many child processes to spawn, or start multiple processes per the concurrency settings but then not pass those settings on because there's no purpose to doing so.

However, we've had this design and implementation for many years, so we can't really reverse course now.  However, perhaps we can find a way to improve the situation as part of the `"version": 2.0` revisions to the custom transfer protocol.

#### Other Issues

As a final note, I think your [comments](https://github.com/git-lfs/git-lfs/pull/6119#issuecomment-3394912888) about the idle timeout and lack of retry handling in the existing custom transfer adapter are worth noting the proposal document as well, if only as concerns to be addressed later.
